src/query-compiler/default-query-compiler.ts

```diff
@@ -1327,6 +1337,53 @@ export class DefaultQueryCompiler
     }
   }
 
+  protected override visitJSONPathReference(node: JSONPathReferenceNode): void {
+    if (node.is$) {
+      this.append(node.operator)
+      this.visitNode(node.jsonPath)
+    } else {
```

Comment:
This feels like a hack: `visitJSONPath` is not called at all if its parent doesn't have `is$: true`. Same for `visitJSONPathLeg`. 

I think we need to have a separate node completely for json paths and postgres-style nested references. The `visit` functions should always be called if there's a node of that kind in the tree.

If you're modelling this after some other similar case in the codebase, we should fix that too (I can do that).

Reply:
I didn't like it either. Glad you noticed that as well.
I don't remember it being inspired by other things in Kysely's tree.
It just got changed a lot over time as I found out new things and new gotchas between the dialects.

Went back to the drawing board with d12e4478ac0c7164cbc51fa0649255b1213c5101.
Yeah, seriously went to a drawing board 😄 
![image](https://github.com/kysely-org/kysely/assets/14938291/18342f20-b2de-4c95-b3ef-3bcb8c2204fa)

Reply:
Would this make all references produce a `JSONReferenceNode`? Or would `ref('foo.bar')` produce just a `ReferenceNode` and `ref('foo.bar', '->').key('baz')` would produce a JSONReferenceNode?

Reply:
Perfect!

Reply:
Currently if `op` is undefined, it produces a  `ReferenceNode`, otherwise it produces a  `JSONReferenceNode`.